app-cpp/appPtr.hpp

```diff
@@ -0,0 +1,436 @@
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            auto refCounts = std::make_pair(mStrongCount, ++mWeakCount);
+            return refCounts;
+        }
+
+        std::pair<std::size_t, std::size_t> DecrementWeak()
+        {
+            std::lock_guard<std::mutex> lock{mMutex};
+            auto refCounts = std::make_pair(mStrongCount, --mWeakCount);
+            return refCounts;
+        }
+
+    private:
+        std::size_t mStrongCount{1}; // Default constructs with initial refcount of 1
+        std::size_t mWeakCount{0};
+        std::mutex mMutex{};
+        T* mData{}; // Used by weak pointers to obtain a new shared pointer
+    };              // class ControlBlock
+
+private:
+    template <typename T>
+    friend class weak_ptr;
+
+    shared_ptr(ControlBlock* inBlock)
+    {
+        mData = inBlock->get();
+        mControl = inBlock;
+        mControl->IncrementStrong();
+    }
+
+private:
+    T* mData{};
+    ControlBlock* mControl{};
+}; // class shared_ptr<T>
+
+// TRICKY: Variatic template arguments 
+// Template that takes any number and type of arguments
+template<typename T, typename ...Args>
+shared_ptr<T> make_unique(Args... inArgs)
+{
+    T* data = new T{std::forward<Args>(inArgs)...};
+    shared_ptr<T> unique{data};
+    return unique;
+}
+
+// WEAK POINTER
+template <typename T>
+class weak_ptr
+{
+public:
+    weak_ptr() : 
+        mControl{nullptr}
+    {
+
+    }
+
+    weak_ptr(shared_ptr<T>& inPtr)
+    {
+        if (inPtr)
+        {
+            mControl = inPtr.mControl;
+            mControl->IncrementWeak();
+        }
+        // Note: ControlBlock::Ctor{} sets strong count to 1
+    }
+
+    ~weak_ptr()
+    {
+        reset();
+    };
+
+    void reset()
+    {
+        if (mControl != nullptr)
+        {
+            auto [strong, weak] = mControl->DecrementWeak();
+            if (strong <= 0)
+            {
+                // delete data not available for weak_ptr
+                if (weak <= 0)
+                {
+                    delete mControl;
+                }
+            }
+        }
+        mControl = nullptr;
+    }
+
+    // RO5 Methods
+    // Shared Assign
+    weak_ptr& operator=(shared_ptr<T>& inShared)
+    {
+        // Copying to self should be a NOP
+        if (this != &inShared)
+        {
+            mControl = inShared.mControl;
+            mControl->IncrementWeak();
+        }
+        return *this;
+    }
+
+    // Move Ctor
+    weak_ptr (weak_ptr&& ioMove)
+    {
+        std::swap(mControl, ioMove.mControl);
+    }
+
+    // Move Operand
+    weak_ptr& operator=(weak_ptr&& ioMove)
+    {
+        // Moving to self should be a NOP
+        if (this != &ioMove)
+        {
+            std::swap(mControl, ioMove.mControl);
+        }
+        return *this;
+    }
+
+    // Copy Ctor
+    weak_ptr (const weak_ptr& inCopy)
+    {
+        mControl = inCopy.mControl;
+        mControl->IncrementStrong();
```

Comment:
Same "test for nullptr (operator bool), then increment, then set mControl" issue here